File: windows/BorderlessMouse/Protocol/ClipboardContent.cs
using System.Buffers.Binary;
using System.Text;
using static BorderlessMouse.Localization.L10n;

namespace BorderlessMouse.Protocol;

public enum ClipboardFormat : byte
{
    Utf8Text = 0,
    Png = 1,
}

/// <summary>Wspólny format schowka; obrazy na łączu są zawsze PNG.</summary>
public sealed class ClipboardContent
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);
    public ClipboardFormat Format { get; }
    public ReadOnlyMemory<byte> Data { get; }
    public string? Text => Format == ClipboardFormat.Utf8Text ? StrictUtf8.GetString(Data.Span) : null;
    public string Summary => Text is { } text
        ? T($"{text.Length} zn.", $"{text.Length} characters")
        : T($"obraz PNG ({(Data.Length + 1023) / 1024} KiB)", $"PNG image ({(Data.Length + 1023) / 1024} KiB)");

    private ClipboardContent(ClipboardFormat format, byte[] data)
    {
        Format = format;
        Data = data;
    }

    public static ClipboardContent? FromText(string text) => Create(ClipboardFormat.Utf8Text, Encoding.UTF8.GetBytes(text));

    public static ClipboardContent? Create(ClipboardFormat format, ReadOnlySpan<byte> data)
    {
        switch (format)
        {
            case ClipboardFormat.Utf8Text:
                if (data.IsEmpty || data.Length > ProtocolConstants.MaxClipboardBytes) return null;
                try { StrictUtf8.GetCharCount(data); }
                catch (DecoderFallbackException) { return null; }
                break;
            case ClipboardFormat.Png:
                // Sprawdź wymiary przed dekodowaniem mocno skompresowanego obrazu.
                if (data.Length < 33 || data.Length > ProtocolConstants.MaxClipboardImageBytes ||
                    !data.StartsWith(new byte[] { 137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82 })) return null;
                var width = (ulong)BinaryPrimitives.ReadUInt32BigEndian(data[16..]);
                var height = (ulong)BinaryPrimitives.ReadUInt32BigEndian(data[20..]);
                if (width == 0 || height == 0 || width * height > ProtocolConstants.MaxClipboardImagePixels) return null;
                break;
            default:
                return null;
        }
        return new ClipboardContent(format, data.ToArray());
    }
}

File: macos/BorderlessMouse/Sources/Protocol/ClipboardContent.swift
import Foundation

/// Wspólny format schowka; obrazy na łączu są zawsze PNG.
struct ClipboardContent: Equatable {
    let format: ClipboardFormat
    let data: Data

    init?(format: ClipboardFormat, data: Data) {
        switch format {
        case .utf8Text:
            guard !data.isEmpty, data.count <= ProtocolConstants.maxClipboardBytes,
                  String(data: data, encoding: .utf8) != nil else { return nil }
        case .png:
            // Sprawdź rozmiary przed dekodowaniem, także dla mocno skompresowanych PNG.
            guard data.count <= ProtocolConstants.maxClipboardImageBytes,
                  data.count >= 33,
                  data.starts(with: [137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82]) else { return nil }
            let header = [UInt8](data.prefix(24))
            let width = header[16..<20].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
            let height = header[20..<24].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
            guard width > 0, height > 0,
                  width * height <= ProtocolConstants.maxClipboardImagePixels else { return nil }
        }
        self.format = format
        self.data = data
    }

    init?(payload: [UInt8]) {
        guard let first = payload.first, let format = ClipboardFormat(rawValue: first) else { return nil }
        self.init(format: format, data: Data(payload.dropFirst()))
    }

    var text: String? {
        format == .utf8Text ? String(data: data, encoding: .utf8) : nil
    }

    var summary: String {
        if let text { return L10n.text("\(text.count) zn.", "\(text.count) characters") }
        return L10n.text("obraz PNG (\((data.count + 1023) / 1024) KiB)",
                         "PNG image (\((data.count + 1023) / 1024) KiB)")
    }
}
